### PR TITLE
Feat: add RBAC support

### PR DESCRIPTION
**Pull request by barnettZQG**
Signed-off-by: barnettZQG <barnett.zqg@gmail.com>

### Description of your changes

Ref: oam-dev/velaux#387

This PR provider a simple R(P)BAC (policy&role based access control) solution for VelaUX, that added the following concepts:

* Role

A role is a description of a set of permission policies. On the other hand, it is also an abstraction of identity. We support two-level roles, including platform role and project role.

example:
```
name: app-developer
alias: App Developer
project: demo
permissions: ["app-manage"]
```

* Permission

A permission policy includes resources, actions, and effect types. The resources are uniformly described by type and entity, which resolved the problem that traditional RBAC could not override data permissions. 

example:
```
name: app1-manage
project: demo
resource: ["project:demo/application:app1/*"]
actions: ["*"]
effect: Allow
principal: {}
condition: {}
```

* PermissionTemplate

The permission policy templates are share resources, we generate templates by default to facilitate the creation of permission policies under each project.

example:
```
name: app-manage
resource: ["project:{projectName}/application:*"]
actions: ["*"]
scope: project
effect: Allow
```

* User bind role

bind platform roles(User model):
```
name: xxx
userRoles: ["admin"]
```
bind project roles(ProjectUser model):
```
userName: xxx
roles: ["admin"]
```

I have:

- [x] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/oam-dev/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

File: docs/examples/rbac/rbac.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+project: demo
+permPolicies: ["app-manage"]
+```
+
+```yaml
+name: admin
+permPolicies: ["all"]
+```
+
+PermPolicy:
```
Comment (wonderflow) [inline docs/examples/rbac/rbac.md:32]:
```suggestion
Permission:
```

Policy  already used in  vela Application, let's avoid confusion.

File: docs/examples/rbac/rbac.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+```yaml
+name: all
+resource: ["*"]
+actions: ["*"]
+effect: Allow
+principal: {}
+condition: {}
+```
+
+PermPolicyTemplate:
```
Comment (wonderflow) [inline docs/examples/rbac/rbac.md:89]:
```suggestion
PermissionTemplate:
```

File: pkg/apiserver/rest/apis/v1/types.go [hunk 3]
```diff
@@ -289,10 +289,10 @@ type ClusterBase struct {
 
 // ListApplicationOptions list application  query options
 type ListApplicationOptions struct {
-	Project    string `json:"project"`
-	Env        string `json:"env"`
-	TargetName string `json:"targetName"`
-	Query      string `json:"query"`
+	Project    []string `json:"project"`
```
Comment (FogDong) [inline pkg/apiserver/rest/apis/v1/types.go:292]:
```suggestion
	Projects    []string `json:"projects"`
```

File: pkg/apiserver/model/user.go [hunk 4]
```diff
@@ -27,8 +27,14 @@ import (
 func init() {
 	RegisterModel(&User{})
 	RegisterModel(&ProjectUser{})
+	RegisterModel(&Role{})
+	RegisterModel(&PermPolicy{})
+	RegisterModel(&PermPolicyTemplate{})
 }
 
+// DefaultAdminUserName default admin user name
+var DefaultAdminUserName = "admin"
```
Comment (FogDong) [inline pkg/apiserver/model/user.go:36]:
```suggestion
const DefaultAdminUserName = "admin"
```

File: pkg/apiserver/rest/usecase/authentication.go [hunk 5]
```diff
@@ -287,8 +290,12 @@ func (a *authenticationUsecaseImpl) GetLoginType(ctx context.Context) (*apisv1.G
 	if err != nil {
 		return nil, err
 	}
+	loginType := sysInfo.LoginType
+	if loginType == "" {
```
Comment (FogDong) [inline pkg/apiserver/rest/usecase/authentication.go:294]: The loginType should be set when systemInfo is initiliazed so we don't need to check here second time.

Comment (barnettZQG) [reply pkg/apiserver/rest/usecase/authentication.go:294]: Yes, if the user upgrades from 1.2, the login type will be empty.

File: pkg/apiserver/rest/usecase/rbac.go [hunk 6]
```diff
@@ -0,0 +1,783 @@
...
+var reg = regexp.MustCompile(`(?U)\{.*\}`)
+
+var defaultProjectPermPolicyTemplate = []*model.PermPolicyTemplate{
+	{
+		Name:      "app-management",
+		Alias:     "App Management",
+		Resources: []string{"project:${projectName}/application:*/*"},
+		Actions:   []string{"*"},
+		Effect:    "Allow",
+		Level:     "project",
```
Comment (FogDong) [inline pkg/apiserver/rest/usecase/rbac.go:49]: Maybe `Scope` will be better of `Level`?

File: pkg/apiserver/rest/usecase/rbac.go [hunk 7]
```diff
@@ -0,0 +1,783 @@
...
+func registerResourceAction(resource string, actions ...string) {
+	lock.Lock()
+	defer lock.Unlock()
+	if resourceActions == nil {
+		resourceActions = make(map[string][]string)
+	}
+	if _, exist := ResourceMaps[resource]; !exist {
+		path, err := checkResourcePath(resource)
+		if err != nil {
+			panic(fmt.Sprintf("resource %s is not exist", resource))
```
Comment (FogDong) [inline pkg/apiserver/rest/usecase/rbac.go:269]: Should we use panic here?

File: pkg/apiserver/rest/usecase/system_info.go [hunk 8]
```diff
@@ -52,6 +51,9 @@ func (u systemInfoUsecaseImpl) GetSystemInfo(ctx context.Context) (*v1.SystemInf
 	}
 	if len(entities) != 0 {
 		info := entities[0].(*model.SystemInfo)
+		if info.LoginType == "" {
```
Comment (FogDong) [inline pkg/apiserver/rest/usecase/system_info.go:54]: same

File: pkg/apiserver/rest/rest_server.go [hunk 9]
```diff
@@ -78,13 +71,14 @@ type leaderConfig struct {
...
 	Run(context.Context) error
-	RegisterServices() restfulspec.Config
+	RegisterServices(ctx context.Context, initDatabase bool) restfulspec.Config
 }
 
 type restServer struct {
 	webContainer *restful.Container
 	cfg          Config
 	dataStore    datastore.DataStore
+	someUsecases map[string]interface{}
```
Comment (FogDong) [inline pkg/apiserver/rest/rest_server.go:81]:
```suggestion
	usecases map[string]interface{}
```

File: pkg/apiserver/rest/usecase/rbac.go (hunk 7 above)
Comment (barnettZQG) [reply pkg/apiserver/rest/usecase/rbac.go:269]: This error represents a coding exception.

File: pkg/apiserver/rest/usecase/rbac.go (hunk 6 above)
Comment (barnettZQG) [reply pkg/apiserver/rest/usecase/rbac.go:49]: agree

File: pkg/apiserver/rest/usecase/user.go [hunk 10]
```diff
@@ -142,12 +203,14 @@ func (u *userUsecaseImpl) CreateUser(ctx context.Context, req apisv1.CreateUserR
 	if err != nil {
 		return nil, err
 	}
+	// TODO: validate the roles, them must all are platform role
```
Comment (wonderflow) [inline pkg/apiserver/rest/usecase/user.go:206]:
```suggestion
	// TODO: validate the roles, they must be platform roles
```

File: pkg/apiserver/rest/utils/bcode/003_project.go [hunk 11]
```diff
@@ -27,3 +27,21 @@ var ErrProjectNamespaceFail = NewBcode(400, 30003, "project bind namespace failu
 
 // ErrProjectNamespaceIsExist the namespace belongs to the other project
 var ErrProjectNamespaceIsExist = NewBcode(400, 30004, "the namespace belongs to the other project")
+
+// ErrProjectDenyDeleteByApplication the project cannot be deleted because there are some applications
+var ErrProjectDenyDeleteByApplication = NewBcode(400, 30005, "the project cannot be deleted because there are some applications")
```
Comment (wonderflow) [inline pkg/apiserver/rest/utils/bcode/003_project.go:32]:
```suggestion
var ErrProjectDenyDeleteByApplication = NewBcode(400, 30005, "the project can't be deleted as there are applications running inside")
```

File: pkg/apiserver/rest/utils/bcode/003_project.go [hunk 12]
```diff
@@ -27,3 +27,21 @@ var ErrProjectNamespaceFail = NewBcode(400, 30003, "project bind namespace failu
...
+
+// ErrProjectDenyDeleteByEnvironment the project cannot be deleted because there are some environments
+var ErrProjectDenyDeleteByEnvironment = NewBcode(400, 30006, "the project cannot be deleted because there are some environments")
```
Comment (wonderflow) [inline pkg/apiserver/rest/utils/bcode/003_project.go:35]:
```suggestion
var ErrProjectDenyDeleteByEnvironment = NewBcode(400, 30006, "the project can't be deleted before you clean up all the environments inside")
```

File: pkg/apiserver/rest/utils/bcode/003_project.go [hunk 13]
```diff
@@ -27,3 +27,21 @@ var ErrProjectNamespaceFail = NewBcode(400, 30003, "project bind namespace failu
...
+
+// ErrProjectDenyDeleteByTarget the project cannot be deleted because there are some targets
+var ErrProjectDenyDeleteByTarget = NewBcode(400, 30007, "the project cannot be deleted because there are some targets")
```
Comment (wonderflow) [inline pkg/apiserver/rest/utils/bcode/003_project.go:38]:
```suggestion
var ErrProjectDenyDeleteByTarget = NewBcode(400, 30007, "the project can't be deleted before you clean up all these targets inside")
```

File: pkg/apiserver/rest/utils/bcode/003_project.go [hunk 14]
```diff
@@ -27,3 +27,21 @@ var ErrProjectNamespaceFail = NewBcode(400, 30003, "project bind namespace failu
...
+
+// ErrProjectRoleCheckFailure means role is not exit or role is not belong to this project.
+var ErrProjectRoleCheckFailure = NewBcode(400, 30008, "the role is not exit or not belong to this project")
```
Comment (wonderflow) [inline pkg/apiserver/rest/utils/bcode/003_project.go:41]:
```suggestion
var ErrProjectRoleCheckFailure = NewBcode(400, 30008, "the specified  role does't belong to this project or not exist")
```

File: pkg/apiserver/rest/utils/bcode/003_project.go [hunk 15]
```diff
@@ -27,3 +27,21 @@ var ErrProjectNamespaceFail = NewBcode(400, 30003, "project bind namespace failu
...
+
+// ErrProjectUserExist means the use is exit in this project
+var ErrProjectUserExist = NewBcode(400, 30009, "the use is exit in this project")
```
Comment (wonderflow) [inline pkg/apiserver/rest/utils/bcode/003_project.go:44]:
```suggestion
var ErrProjectUserExist = NewBcode(400, 30009, "the user is already exist in this project")
```

File: pkg/apiserver/rest/utils/bcode/003_project.go [hunk 16]
```diff
@@ -27,3 +27,21 @@ var ErrProjectNamespaceFail = NewBcode(400, 30003, "project bind namespace failu
...
+
+// ErrProjectOwnerIsNotExist means the owner is invalid
+var ErrProjectOwnerIsNotExist = NewBcode(400, 30010, "the project owner is invalid")
```
Comment (wonderflow) [inline pkg/apiserver/rest/utils/bcode/003_project.go:47]:
```suggestion
var ErrProjectOwnerIsNotExist = NewBcode(400, 30010, "the project owner name is invalid")
```

**Review by wonderflow — APPROVED** [review]

**Review by Somefive — APPROVED** [review]

**Review by FogDong — APPROVED** [review]

**Review by yangsoon — APPROVED** [review]

**Pull request merged**